### PR TITLE
Implements Direct State Access and Multi Bind

### PR DESCRIPTION
**Pull request by mpnico**
### Implements OpenGL 4.4 Multi Bind and OpenGL 4.5 DSA
The goal of this PR is to reduce GL calls by using OpenGL 4.4 and 4.5 new extensions (`ARB_multi_bind` and `ARB_direct_state_access`), closing #2508.

Multi bind allows to bind multiple buffers in one call whereas DSA allows to change states of objects without binding them (buffers / textures, framebuffers, …).

Unfortunately, I didn't see major performance increase during my tests, do not expect more than a few %.

About VertexArray: the current implementation is not 100% full DSA. The "DSA" version of `glEnableVertexAttribArray` / `glDisableVertexAttribArray` are not used because of a big negative performance impact (around 15% on SMO).

About Framebuffer: as stated by @ReinUsesLisp in #2508, I encountered a bug with `CreateFramebuffers` so I made an exception only for NVidia.

### What to test
Basically every games for potential regressions. My system is NVidia only so AMD/Intel tests are needed. 

During my tests, I also encountered a strange flickering bug in Monster Hunter Rise which disappears when activating then disabling OpenGL log level. I assume it is a driver bug with my system but this game should be tested.

### CODE REVIEW

**Comment by A-w-x** [comment]
intel linux

links awakening
[Ryujinx_1.0.0+be5dffc_2021-08-16_06-05-29.log](https://github.com/Ryujinx/Ryujinx/files/6992031/Ryujinx_1.0.0%2Bbe5dffc_2021-08-16_06-05-29.log)
Screenshot_2021-08-16_06-09-52

File: Ryujinx.Graphics.OpenGL/Pipeline.cs [hunk 1]
```diff
@@ -1006,6 +994,7 @@ public void SetTransformFeedbackBuffers(ReadOnlySpan<BufferRange> buffers)
             }
 
             int count = Math.Min(buffers.Length, Constants.MaxTransformFeedbackBuffers);
+            int[] bufferHandles = new int[count];
```
Comment (riperiperi) [inline Ryujinx.Graphics.OpenGL/Pipeline.cs:997]: Use `stackalloc` to avoid allocating arrays in the heap each time this is called.

File: Ryujinx.Graphics.OpenGL/Pipeline.cs [hunk 2]
```diff
@@ -1118,18 +1109,29 @@ private void SetBuffers(int first, ReadOnlySpan<BufferRange> buffers, bool isSto
         {
             BufferRangeTarget target = isStorage ? BufferRangeTarget.ShaderStorageBuffer : BufferRangeTarget.UniformBuffer;
 
+            int[] bufferHandles = new int[buffers.Length];
+            IntPtr[] bufferOffsets = new IntPtr[buffers.Length];
+            IntPtr[] bufferSizes = new IntPtr[buffers.Length];
```
Comment (riperiperi) [inline Ryujinx.Graphics.OpenGL/Pipeline.cs:1114]: `stackalloc` these - SetBuffers is called a lot and allocating arrays each time isn't great.

File: Ryujinx.Graphics.OpenGL/Framebuffer.cs [hunk 3]
```diff
@@ -17,9 +17,28 @@ class Framebuffer : IDisposable
...
+            {
+                // Nvidia bug, cf. https://forums.developer.nvidia.com/t/bug-spec-violation-checknamedframebufferstatus-returns-gl-framebuffer-incomplete-dimensions-ext-under-gl-4-5-core/58647
+                GL.GenFramebuffers(1, &localHandle);
+                GL.BindFramebuffer(FramebufferTarget.Framebuffer, localHandle);
+            }
+            else
+            {
+                GL.CreateFramebuffers(1, &localHandle);
+            }
+
```
Comment (riperiperi) [inline Ryujinx.Graphics.OpenGL/Framebuffer.cs:34]: Double newline.

File: Ryujinx.Graphics.OpenGL/Image/TextureView.cs [hunk 4]
```diff
@@ -276,8 +253,12 @@ private void ReadFrom2D(IntPtr data, int layer, int level, int width, int height
             TextureTarget target = Target.Convert();
 
             int mipSize = Info.GetMipSize2D(level);
+            int maxLevels = 1 + (int)Math.Floor(Math.Log2(width * height));
```
Comment (riperiperi) [inline Ryujinx.Graphics.OpenGL/Image/TextureView.cs:256]: This calculation doesn't make any sense. Given a 256x256 texture, this will calculate 17 max levels... which is pretty far off.

```
256
128
64
32
16
8
4
2
1
```
There are 9 mip levels in this situation. With your current calculation, the correct result is obtained with 256x1.

File: Ryujinx.Graphics.OpenGL/VertexArray.cs [hunk 5]
```diff
@@ -20,9 +20,18 @@ class VertexArray : IDisposable
         private uint _vertexAttribsInUse;
         private uint _vertexBuffersInUse;
 
+        public unsafe int Create()
+        {
+            int localHnadle;
```
Comment (riperiperi) [inline Ryujinx.Graphics.OpenGL/VertexArray.cs:25]: `localHandle`

**Review by riperiperi — COMMENTED** [review]
Just a quick initial review. I've yet to do more testing, but I can confirm the Link's Awakening issue also happens on NVIDIA.
image

File: Ryujinx.Graphics.OpenGL/Image/TextureView.cs (hunk 4 above)
Comment (mpnico) [reply Ryujinx.Graphics.OpenGL/Image/TextureView.cs:256]: Well this is a commit mistake. This was a test to avoid flood of OpenGL errors in logs in SMO which happens on master too. No relation with DSA so I will remove it.

**Comment by mpnico** [comment]
Fixed Link's Awakening and applied code review changes.

**Comment by mpnico** [comment]
I found a workaround for MHR flickering. I also pushed a few more fixes

File: Ryujinx.Graphics.OpenGL/Buffer.cs [hunk 6]
```diff
@@ -8,21 +8,18 @@ static class Buffer
...
+            valueArr[0] = value;
+
+            GL.ClearNamedBufferSubData<uint>(
+                destination.ToInt32(),
+                PixelInternalFormat.Rgba8ui,
+                (IntPtr)offset,
+                size,
+                PixelFormat.RgbaInteger,
+                PixelType.UnsignedByte,
+                valueArr.ToArray());
```
Comment (gdkchan) [inline Ryujinx.Graphics.OpenGL/Buffer.cs:22]: Calling `ToArray` here negates the benefits of using the stackalloc (as you will allocate a new array on the heap anyway).

File: Ryujinx.Graphics.Gpu/Image/Texture.cs [hunk 7]
```diff
@@ -833,7 +833,7 @@ public void Flush(bool tracked = true)
         /// </summary>
         public void ExternalFlush(ulong address, ulong size)
         {
-            if (!IsModified)
+            if (!IsModified || HostTexture.isDisposed())
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/Texture.cs:836]: There is no need for this `isDisposed` method (btw the correct name would be `IsDisposed` as funciton names always start with upper case). You can just have a bool field and set it to true when `DisposeTextures` is called. Then here it would just check `if (!IsModified || _disposed)`.

Comment (gdkchan) [reply Ryujinx.Graphics.Gpu/Image/Texture.cs:836]: Also I'm not sure of the effects of leaving `IsModified` set to true. Maybe it could cause the texture to be flushed all the time. @riperiperi should be able to give a better suggestion here.

File: Ryujinx.Graphics.OpenGL/Buffer.cs (hunk 6 above)
Comment (mpnico) [reply Ryujinx.Graphics.OpenGL/Buffer.cs:22]: This is the fix for Monster Hunter Rise. Maybe using `MemoryMarshal.GetReference` is better ?

Comment (gdkchan) [reply Ryujinx.Graphics.OpenGL/Buffer.cs:22]: If theres an overload that takes ref, you can just pass `ref valueArr[0]`. Or well, just `ref value` since theres only one value.

Comment (mpnico) [reply Ryujinx.Graphics.OpenGL/Buffer.cs:22]: Method support to pass `ref value` but does not fix the bug that way. `MemoryMarshal.GetReference` does fix it but I don't know it efficiency.

Comment (gdkchan) [reply Ryujinx.Graphics.OpenGL/Buffer.cs:22]: I think passing the array is just fixing the issue by chance. It's possible that the driver is reading past the pointer that we pass. Try changing the stackalloc size from 1 to 4, and passing the pointer directly, see if that fixes it.

File: Ryujinx.Graphics.Gpu/Image/Texture.cs (hunk 7 above)
Comment (mpnico) [reply Ryujinx.Graphics.Gpu/Image/Texture.cs:836]: I did some tests and, at least in MK8D where I found those disposed textures, and no excessive flush happen so I will make the well advised changes and won't touch `IsModified`

File: Ryujinx.Graphics.OpenGL/Buffer.cs (hunk 6 above)
Comment (mpnico) [reply Ryujinx.Graphics.OpenGL/Buffer.cs:22]: Good idea but unfortunately it doesn't fix the bug. But you gave me the idea to change `PixelInternalFormat` to a 1 component only (`GL_R32UI`) format and it fixes the bug. I don't know if it's very less efficient though.

Comment (mpnico) [reply Ryujinx.Graphics.OpenGL/Buffer.cs]: Finally the best fix is using `MemoryMarshal.GetReference` or not using DSA here. I choose the first one.

**Comment by mpnico** [comment]
Rebased and fixed a problem with some UE games. It need some more tests on AMD as I also pushed a potential fix for garbage pixels at edges of the window.

**Comment by TSRBerry** [comment]
Needs a rebase

**Comment by marysaka** [comment]
Closing as stale.
